public/app/plugins/datasource/tempo/traceql/autocomplete.ts

```diff
@@ -0,0 +1,235 @@
+   * @param situation
+   * @private
+   */
+  private async getCompletions(situation: Situation): Promise<Completion[]> {
+    if (!Object.keys(this.tags).length) {
+      return [];
+    }
+    switch (situation.type) {
+      // Not really sure what would make sense to suggest in this case so just leave it
+      case 'UNKNOWN': {
+        return [];
+      }
+      case 'EMPTY': {
+        return Object.keys(this.tags).map((key) => {
+          return {
+            label: key,
+            insertText: `{${key}="`,
+            type: 'TAG_NAME',
+          };
+        });
+      }
+      case 'IN_TAG_NAME':
+        return Object.keys(this.tags).map((key) => {
+          return {
+            label: key,
+            insertText: key,
+            type: 'TAG_NAME',
+          };
+        });
+      case 'IN_TAG_VALUE':
+        return await this.languageProvider.getOptions(situation.tagName).then((res) => {
+          const items: Completion[] = [];
+          res.forEach((val) => {
+            if (val?.label) {
+              items.push({
+                label: val.label,
+                insertText: situation.betweenQuotes ? val.label : `"${val.label}"`,
+                type: 'TAG_VALUE',
+              });
+            }
+          });
+          return items;
+        });
+      default:
+        throw new Error(`Unexpected situation ${situation}`);
+    }
+  }
+}
+
+/**
+ * Get item kind which is used for icon next to the suggestion.
+ * @param type
+ * @param monaco
+ */
+function getMonacoCompletionItemKind(type: CompletionType, monaco: Monaco): monacoTypes.languages.CompletionItemKind {
+  switch (type) {
+    case 'TAG_NAME':
+      return monaco.languages.CompletionItemKind.Enum;
+    case 'TAG_VALUE':
+      return monaco.languages.CompletionItemKind.EnumMember;
+    default:
+      throw new Error(`Unexpected CompletionType: ${type}`);
+  }
+}
+
+export type CompletionType = 'TAG_NAME' | 'TAG_VALUE';
+type Completion = {
+  type: CompletionType;
+  label: string;
+  insertText: string;
+};
+
+export type Tag = {
+  name: string;
+  value: string;
+};
+
+export type Situation =
+  | {
+      type: 'UNKNOWN';
+    }
+  | {
+      type: 'EMPTY';
+    }
+  | {
+      type: 'IN_TAG_NAME';
+      otherTags: Tag[];
+    }
+  | {
+      type: 'IN_TAG_VALUE';
+      tagName: string;
+      betweenQuotes: boolean;
+      otherTags: Tag[];
+    };
+
+/**
+ * Figure out where is the cursor and what kind of suggestions are appropriate.
+ * As currently TraceQL handles just a simple {foo="bar", baz="zyx"} kind of values we can do with simple regex to figure
+ * out where we are with the cursor.
+ * @param text
+ * @param offset
+ */
+function getSituation(text: string, offset: number): Situation {
+  if (text === '') {
+    return {
+      type: 'EMPTY',
+    };
+  }
+
+  // Get all the tags so far in the query so we can do some more filtering.
+  const matches = text.matchAll(/(\w+)="(\w+)"/g);
```

Comment:
Btw depending on the how we approach parsing these should also support other operators than `=` ie `<=` `!=` etc. I have to also check in my datasource whether that should be supported or not.